### PR TITLE
Refactor mesh exporter

### PR DESCRIPTION
**Pull request by Jason0214**
intend to solve #46

### CODE REVIEW

**Comment by Jason0214** [comment]
I am thinking about adding a test case which has bone-weight-zero vertex

File: io_scene_godot/converters/mesh.py [hunk 1]
```diff
@@ -219,156 +102,249 @@ def validate_morph_mesh_modifiers(mesh_object):
...
+        mesh = self.object.to_mesh(bpy.context.scene,
+                                   True,
+                                   "RENDER")
+
+        # Prepare the mesh for export
         triangulate_mesh(mesh)
 
-        if has_tangents:
+        uv_layer_count = len(mesh.uv_textures)
+        if uv_layer_count > 2:
```
Comment (sdfgeoff) [inline io_scene_godot/converters/mesh.py]: I think this was copied over from before, but using the `min` function would be prettier

File: io_scene_godot/converters/mesh.py [hunk 2]
```diff
@@ -105,85 +67,6 @@ def triangulate_mesh(mesh):
...
-    bpy.data.meshes.remove(mesh)
-
-    if armature_data is not None:
-        armature_data.pose_position = original_pose_position
-        bpy.context.scene.update()
-
-    return surfaces
-
-
 def extract_shape_keys(blender_shape_keys):
```
Comment (sdfgeoff) [inline io_scene_godot/converters/mesh.py]: I think that this and `initialize_surfaces_morph_data` should be `@staticmethod`'s of the `MeshResourceExporter` class

File: tests/export_test_scenes.py [hunk 3]
```diff
@@ -10,7 +10,10 @@
 def export_escn(out_file):
     """Fake the export operator call"""
     import io_scene_godot
-    io_scene_godot.export(out_file, {})
+    try:
```
[the remark this answers is not in the document]
Comment (Jason0214) [reply tests/export_test_scenes.py]: Ah, I add a test case of a rig with no bone weight vertex, I wouldn't want the makefile to abort, as the expected test result would be raise an error and that rig-with-no-bone-weight-vertex not appear in godot-project/exports

File: io_scene_godot/structures.py [hunk 4]
```diff
@@ -272,6 +272,23 @@ def to_string(self):
...
+    with to_string() method"""
+    def __init__(self):
+        super().__init__()
+
+        self.__str__ = self.to_string
+
+    def to_string(self):
+        """Convert the map to serialized form"""
+        return ("{\n\t" +
+                '\n\t'.join(['"{}": {},'.format(k, to_string(v))
```
[the remark this answers is not in the document]
Comment (Jason0214) [reply io_scene_godot/structures.py]: ok, I will change it

File: io_scene_godot/converters/mesh.py [hunk 5]
```diff
@@ -219,156 +102,249 @@ def validate_morph_mesh_modifiers(mesh_object):
...
+def fix_vertex(vtx):
+    """Changes a single position vector from y-up to z-up"""
+    return mathutils.Vector((vtx.x, vtx.z, -vtx.y))
+
+
+def get_modifier_armature_data(mesh_object):
+    """Get the armature modifier of a blender object
+    if does not have one, return None"""
+    for modifier in mesh_object.modifiers:
+        if isinstance(modifier, bpy.types.ArmatureModifier):
```
[the remark this answers is not in the document]
Comment (Jason0214) [reply io_scene_godot/converters/mesh.py]: Actually I moved this segment of code from armature.py, so you are suggesting it should remain placed there?  My points is this check more related to mesh

File: io_scene_godot/converters/mesh.py [hunk 6]
```diff
@@ -32,26 +32,21 @@ def export_mesh_node(escn_file, export_settings, node, parent_gd_node):
         return parent_gd_node
 
     else:
-        armature_data = armature.get_armature_data(node)
+        mesh_node = NodeTemplate(node.name, "MeshInstance", parent_gd_node)
+        mesh_exporter = MeshResourceExporter(node)
 
-        skeleton_node = None
+        armature_data = get_modifier_armature_data(node)
```
Comment (sdfgeoff) [inline io_scene_godot/converters/mesh.py]: Any reason you don't store the skeleton from the init_mesh_bones_data function?

File: io_scene_godot/converters/mesh.py [hunk 7]
```diff
@@ -105,270 +67,285 @@ def triangulate_mesh(mesh):
...
-    compatible with shape key"""
-    # this black list is not complete
-    modifiers_not_supported = (
-        bpy.types.SubsurfModifier,
-    )
-    for modifier in mesh_object.modifiers:
-        if isinstance(modifier, modifiers_not_supported):
-            return False
-    return True
+        shape_keys_to_export = MeshResourceExporter.extract_shape_keys(
```
Comment (sdfgeoff) [inline io_scene_godot/converters/mesh.py]: You can use `self.extract_shape_keys` which means changing the class name won't break things.

File: io_scene_godot/converters/mesh.py [hunk 8]
```diff
@@ -573,6 +543,47 @@ def get_tup(self):
 
         return tup
 
+    @staticmethod
```
Comment (sdfgeoff) [inline io_scene_godot/converters/mesh.py]: If you use `@classmethod`, then the first parameter (typically called `cls` is a reference to the uninstantiated class. This means you can do `new_vert = cls()`

**Comment by sdfgeoff** [comment]
I'm happy with the structure, so just fix up those minor points above and squash, and I'll merge tomorrow.

File: io_scene_godot/converters/mesh.py (hunk 6 above)
Comment (Jason0214) [reply io_scene_godot/converters/mesh.py]: skeleton is stored in mesh node, not in the ArrayMesh resource, so I place it outside the class

File: tests/export_test_scenes.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (sdfgeoff) [reply tests/export_test_scenes.py]: Hmm, but non-intentional errors will also be caught by this. Normally a testing framework has an idea of if an error is expected, but I can't see any easy way to implement that here.

Maybe you could catch ValidationError explicitly?

File: io_scene_godot/converters/mesh.py (hunk 6 above)
Comment (sdfgeoff) [reply io_scene_godot/converters/mesh.py]: Ok, fair enough.

File: tests/export_test_scenes.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (Jason0214) [reply tests/export_test_scenes.py]: yeah, catch Exception do cause some problems.    I think I will remove the new added test case, it is not so useful

**Comment by Jason0214** [comment]
removed catch `Exception` in `export_test_scene`, it can be reviewed again @sdfgeoff
